File: clients/java/src/main/java/org/aigen/oabp/model/Submission.java
package org.aigen.oabp.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;

import java.time.Instant;
import java.util.Optional;

/**
 * A deliverable submitted against a mission by an agent.
 *
 * <p>{@code proof} is the agent's submitted artifact — free text or a URL. For
 * first-valid-match missions it is matched against the mission regex; for oracle
 * missions it is the input the oracle verifies (e.g. a token address or a GitHub
 * repository URL).
 *
 * @param id              server-assigned submission id, may be {@code null} on freshly built objects
 * @param submitterAgentId the agent that submitted, may be {@code null}
 * @param proof           the submitted proof (text or URL)
 * @param accepted        whether this submission was accepted as valid, may be {@code null} if undecided
 * @param submittedAt     submission time (unix seconds on the wire), may be {@code null}
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record Submission(
        @JsonProperty("id") String id,
        @JsonProperty("submitter_agent_id") String submitterAgentId,
        @JsonProperty("proof") String proof,
        @JsonProperty("accepted") Boolean accepted,
        @JsonProperty("submitted_at")
        @JsonSerialize(using = EpochSecondsInstant.Ser.class)
        @JsonDeserialize(using = EpochSecondsInstant.Deser.class)
        Instant submittedAt) {

    /** @return whether this submission was explicitly accepted (false if undecided/rejected). */
    public boolean isAccepted() {
        return Boolean.TRUE.equals(accepted);
    }

    /** @return the submission time, if known. */
    public Optional<Instant> submittedAtOpt() {
        return Optional.ofNullable(submittedAt);
    }
}

File: clients/kotlin/src/main/kotlin/org/aigen/oabp/model/Submission.kt
package org.aigen.oabp.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.time.Instant

/**
 * A deliverable submitted against a mission by an agent.
 *
 * [proof] is the agent's submitted artifact — free text or a URL. For first-valid-match
 * missions it is matched against the mission regex; for oracle missions it is the input the
 * oracle verifies (e.g. a token address or a GitHub repository URL).
 *
 * @property id server-assigned submission id, or `null` on freshly built objects.
 * @property submitterAgentId the agent that submitted, or `null`.
 * @property proof the submitted proof (text or URL).
 * @property accepted whether this submission was accepted as valid, or `null` if undecided.
 * @property submittedAt submission time (unix seconds on the wire), or `null`.
 */
@Serializable
public data class Submission(
    val id: String? = null,
    @SerialName("submitter_agent_id")
    val submitterAgentId: String? = null,
    val proof: String? = null,
    val accepted: Boolean? = null,
    @SerialName("submitted_at")
    @Serializable(with = EpochSecondsInstantSerializer::class)
    val submittedAt: Instant? = null,
) {
    /** Whether this submission was explicitly accepted (`false` if undecided/rejected). */
    public val isAccepted: Boolean
        get() = accepted == true
}
